### PR TITLE
Refactor `@round` to transform `a+b` to `+(a, b, RoundDown)`

### PR DESCRIPTION
**Pull request by dpsanders**
Transforms e.g.
```
@round(a.lo+b.lo, a.hi+b.hi)
```
to
```
Interval(+(a.lo, b.lo, RoundDown), +(a.hi, b.hi, RoundUp))
```

### CODE REVIEW

**Comment by lbenet** [comment]
Tests are not passing...

**Comment by dpsanders** [comment]
Yes, unfortunately that will require quite a lot of work. But I think this is a good direction to go in.

**Comment by dpsanders** [comment]
Tests are passing on 0.4 and 0.5, and almost on 0.6 with known issues (one in FixedSizeArrays.jl, and the other in Base). cc @lbenet

**Comment by lbenet** [comment]
I'm begining to review this (*huge*) PR... Is it ready to be merged, or the idea is to solve the incompatibilities with Julia v 0.6?

**Comment by dpsanders** [comment]
It's ready to be merged. The only remaining incompatibility is the problem with FixedSizeArrays.jl; I'll see what happens if I change to StaticArrays (in a separate PR).

Sorry about the size of this, but the rounding is so fundamental that it touches almost everything!

**Comment by dpsanders** [comment]
Most of the changes are trivial, though, e.g.
`@round(T, a, b)` loses the `T` to become `@round(a, b)`.

**Comment by dpsanders** [comment]
No problem if you don't feel like reviewing it, though.

File: src/intervals/set_operations.jl [hunk 1]
```diff
@@ -56,18 +56,18 @@ function intersect{T}(a::Interval{T}, b::Interval{T})
...
 
-Returns the "convex hull" of the intervals `a` and `b`, considered as
-(extended) sets of real numbers. That is, the minimum set that contains
-all points in `a` and `b`.
+Returns the "interval hull" of the intervals `a` and `b`, considered as
+(extended) sets of real numbers, i.e. the smallest interval that contains
+all of `a` and `b`.
 """
-hull{T}(a::Interval{T}, b::Interval{T}) = Interval(min(a.lo, b.lo), max(a.hi, b.hi))
+hull{T}(a::Interval{T}, b::Interval{T}) = Interval{T}(min(a.lo, b.lo), max(a.hi, b.hi))
```
Comment (lbenet) [inline src/intervals/set_operations.jl]: I just notice that this may be too restrictive, so we need other methods when `eltype(a)` and `eltype(b)` are different. The following yields an error (also in master): `hull(Interval(1//2,3//1), Interval(1,2))`. The same applies to `union`.

File: src/intervals/trigonometric.jl [hunk 2]
```diff
@@ -44,23 +44,25 @@ function sin{T}(a::Interval{T})
...
+        return @round(sin(a.hi), sin(a.lo)) # Interval(sin(a.hi, RoundDown), sin(a.lo, RoundUp))
 
     elseif ( lo_quadrant == 0 || lo_quadrant==3 ) && ( hi_quadrant==1 || hi_quadrant==2 )
-        return Interval(min(sin(a.lo, RoundDown), sin(a.hi, RoundDown)), one(T))
+        return @round(min(sin(a.lo), sin(a.hi)), 1)
+        # Interval(min(sin(a.lo, RoundDown), sin(a.hi, RoundDown)), one(T))
 
     elseif ( lo_quadrant == 1 || lo_quadrant==2 ) && ( hi_quadrant==3 || hi_quadrant==0 )
-        return Interval(-one(T), max(sin(a.lo, RoundUp), sin(a.hi, RoundUp)))
+        return @round(-1, max(sin(a.lo), sin(a.hi)))
```
Comment (lbenet) [inline src/intervals/trigonometric.jl]: Same comment about conversion; I prefer to use `one(T)` here...

File: src/intervals/trigonometric.jl [hunk 3]
```diff
@@ -86,21 +88,21 @@ function cos{T}(a::Interval{T})
...
-        return Interval(cos(a.lo, RoundDown), cos(a.hi, RoundUp))
+        return @round(cos(a.lo), cos(a.hi))
 
     elseif lo_quadrant == 0 && hi_quadrant==1
-        return Interval(cos(a.hi, RoundDown), cos(a.lo, RoundUp))
+        return @round(cos(a.hi), cos(a.lo))
 
     elseif ( lo_quadrant == 2 || lo_quadrant==3 ) && ( hi_quadrant==0 || hi_quadrant==1 )
-        return Interval(min(cos(a.lo, RoundDown), cos(a.hi, RoundDown)), one(T))
+        return @round(min(cos(a.lo), cos(a.hi)), 1)
```
Comment (lbenet) [inline src/intervals/trigonometric.jl]: Same comment about use of `1` instead of `one(T)`.

File: src/intervals/trigonometric.jl [hunk 4]
```diff
@@ -86,21 +88,21 @@ function cos{T}(a::Interval{T})
...
 
     elseif ( lo_quadrant == 0 || lo_quadrant==1 ) && ( hi_quadrant==2 || hi_quadrant==3 )
-        return Interval(-one(T), max(cos(a.lo, RoundUp), cos(a.hi, RoundUp)))
+        return @round(-1, max(cos(a.lo), cos(a.hi)))
```
Comment (lbenet) [inline src/intervals/trigonometric.jl]: Again...

**Comment by lbenet** [comment]
I just noticed that tests are failing again.

**Comment by lbenet** [comment]
The failing tests seem to be related to a change a suggest in `^(Interval{BigFloat},x::AbstractFloat)`... I hope my comment helps.

**Comment by dpsanders** [comment]
Many thanks for the detailed review! I'll work through your comments.

File: src/intervals/trigonometric.jl [hunk 5]
```diff
@@ -27,15 +27,15 @@ end
 function sin{T}(a::Interval{T})
     isempty(a) && return a
 
-    whole_range = Interval(-one(T), one(T))
+    whole_range = Interval{T}(-1, 1)
```
Comment (dpsanders) [inline src/intervals/trigonometric.jl]: My original motivation for these changes was that I was suspecting that they were introducing a type instability problem (which I doubt, but I just wanted to make sure).

I also think that the `Interval{T}(-1, 1)` is easier to read.

However, now that you pointed it out, I went digging, and found that 
`@edit one(Float64)`
gives the following definition:
```
one{T<:Number}(::Type{T}) = convert(T,1)
```
So it's doing the conversion anyway ;)

Also if you define 
```
f{T}(::Type{T}) = Interval(-one(T), one(T))
g{T}(::Type{T}) = Interval{T}(-1, 1)
```
and do `@code_native f(Float64)` and similarly for `g`, you'll find that there are fewer instructions for `g`. So all round, the new version seems to be a win!

File: src/intervals/trigonometric.jl (hunk 2 above)
Comment (dpsanders) [inline src/intervals/trigonometric.jl]: Here the reason is actually different to change this to not use `one(T)`, namely that it was causing problems with the new version of the macro. I now see that these should be solvable; nonetheless, I actually find it much more readable like this, and the same comment about efficiency (that the conversion still happens anyway) applies.

File: src/intervals/arithmetic.jl [hunk 6]
```diff
@@ -165,25 +184,29 @@ function fma{T}(a::Interval{T}, b::Interval{T}, c::Interval{T})
...
+
     end
 
     lo = setrounding(T, RoundDown) do
         lo1 = fma(a.lo, b.lo, c.lo)
         lo2 = fma(a.lo, b.hi, c.lo)
         lo3 = fma(a.hi, b.lo, c.lo)
         lo4 = fma(a.hi, b.hi, c.lo)
-        min(lo1, lo2, lo3, lo4)
+        min_ignore_nans(lo1, lo2, lo3, lo4)
```
Comment (dpsanders) [inline src/intervals/arithmetic.jl]: That is because that is parsed as `(nai() * 1)..2`. When you write it as `nai() * (1..2)`, it does correctly give `[nan, nan]`.  By the way, I'm not sure why `nan` is output instead of `NaN`?

Comment (dpsanders) [inline src/intervals/arithmetic.jl]: By the way, maybe it is wrong that `Interval` allows to do `Interval(NaN, 2)` at all?

File: src/intervals/functions.jl [hunk 7]
```diff
@@ -18,68 +18,71 @@ function ^(a::Interval{BigFloat}, n::Integer)
...
+                return @round(a.lo^n, a.hi^n)
             else
-                return @round(T, mag(a)^n, mig(a)^n)
+                return @round(mag(a)^n, mig(a)^n)
             end
         end
     end
 end
 
 function sqr{T<:Real}(a::Interval{T})
```
Comment (dpsanders) [inline src/intervals/functions.jl]: The difference is almost negligible (a couple of comparisons), and it just duplicates code unnecessarily.

File: src/intervals/functions.jl [hunk 8]
```diff
@@ -18,68 +18,71 @@ function ^(a::Interval{BigFloat}, n::Integer)
...
+    #     return @round(a.lo^2, a.hi^2)
+    #
+    # elseif a.hi ≤ zero(T)
+    #     return @round(a.hi^2, a.lo^2)
+    # end
+    #
+    # return @round(mig(a)^2, mag(a)^2)
 end
 
+^(a::Interval{BigFloat}, x::AbstractFloat) = a^big(x)
```
Comment (dpsanders) [inline src/intervals/functions.jl]: I don't think it's really necessary. If you want to be careful you should anyway be doing `a^(1//3)` or something like that. I'm actually beginning to wonder if we should be "guessing" the user's intentions anyway. (There was a thread on discourse or somewhere about this recently.)

File: src/intervals/set_operations.jl (hunk 1 above)
Comment (dpsanders) [inline src/intervals/set_operations.jl]: I doubt that in an actual application you would be mixing interval types like that. But we can certainly add `hull(a, b) = hull(promote(a, b)...)`.

File: src/intervals/trigonometric.jl (hunk 5 above)
Comment (lbenet) [inline src/intervals/trigonometric.jl]: I partially agree with your argument 😄. One the one hand, you are right, the code is more readable, the conversion is also there, and calling `one` produces more code. Yet, I can't forget the usual example of type instability where the *extra* conversion saves the day in terms of performance. In any case, maybe for `Interval` the whole discussion is somewhat irrelevant, sorry!, in the sense that we `Interval{Int} is converted by using `float()`. So let's leave it as it is now, and forget my comments on this.

File: src/intervals/arithmetic.jl (hunk 6 above)
Comment (lbenet) [inline src/intervals/arithmetic.jl]: I agree; I didn't notice the parsing issue before...

File: src/intervals/functions.jl (hunk 8 above)
Comment (lbenet) [inline src/intervals/functions.jl]: Yet, our tests are failing right here: the test checks that `x^(1//3) --> x^(1/3)` holds, for `x = biginterval(27)`; the correct answer *must* contain 3. So this is the method that is being called for the RHS. 

Checking this test, `big(1//3)` behaves correctly (renders `BigInt(1)//BigInt(3)`), and the answer contains `3`. However, `big(1/3)` obviously behaves incorrectly (converts the `Float64` representation of `1/3` to `BigFloat`, which is far away from 1/3!).  Incidentally, my proposal is not working either for this.

File: src/intervals/set_operations.jl (hunk 1 above)
Comment (lbenet) [inline src/intervals/set_operations.jl]: I think we should add such a method, as well as the corresponding for `union`. Indeed, it maybe rare in applications...

Comment (dpsanders) [inline src/intervals/set_operations.jl]: Fallback added.

File: src/intervals/trigonometric.jl (hunk 5 above)
Comment (dpsanders) [inline src/intervals/trigonometric.jl]: You were right to raise the issue (and I knew you would when I made the change ;) ). It led to a useful discussion. But I suggest that we leave the new version.

**Comment by dpsanders** [comment]
I split out the rounding macros into a new file, `rounding_macros.jl`. 
(Actually there is now only one rounding macro.)

**Comment by dpsanders** [comment]
I should say that part of the motivation for this whole change, which was in my head but not necessarily mentioned anywhere, was that this will make it much easier to experiment with different types of rounding: the only place that needs to be changed are now the definitions in `rounding.jl`!

File: src/intervals/functions.jl (hunk 8 above)
Comment (dpsanders) [inline src/intervals/functions.jl]: I can imagine the possibility that you would explicitly want to see what happened with an the floating point number "1/3". The alternative is to convert the exponent to an `Interval`.

Comment (dpsanders) [inline src/intervals/functions.jl]: For now I just removed that test and everything now seems to be passing. We could just record this as an issue and leave it for now.

**Comment by dpsanders** [comment]
Tests are passing on 0.4 and 0.5. I think this is ready to merge!

**Comment by dpsanders** [comment]
The macro documentation is in `rounding_macros.jl` -- see https://github.com/dpsanders/ValidatedNumerics.jl/blob/7e67cbd3cbf3b895a29c8839221f3f9e0bb7ff3e/src/intervals/rounding_macros.jl

Is that what you were looking for? 

Since `1/3` is not the true one third, `x^(1/3)` doesn't need to contain the "true" answer for the true one-third power. 

I see this as analogous to the fact that you can do `Interval(1/3)` if you want to use the floating-point number closest to the true one third, or use `@interval(1/3)` if you want to include the true one third.

Similarly you can do `x^(1/3)` to use the floating-point value `1/3`, or e.g. `x^(1//3)` to use the true real number one third.

**Comment by dpsanders** [comment]
I think we should merge this and can come back and change things later if necessary.

**Comment by lbenet** [comment]
Ok. I'm merging.

**Comment by dpsanders** [comment]
Thanks!

File: src/intervals/arithmetic.jl [hunk 9]
```diff
@@ -155,6 +155,25 @@ end
 
 //(a::Interval, b::Interval) = a / b    # to deal with rationals
 
+if VERSION >= v"0.6.0-dev"
```
Comment (tkelman) [inline src/intervals/arithmetic.jl]: this won't work for all 0.6-dev versions, should be more specific about version-depedent cutoffs or use `isdefined`

File: src/intervals/arithmetic.jl [hunk 10]
```diff
@@ -155,6 +155,25 @@ end
...
+    const filter = Iterators.filter
+end
+
+if VERSION < v"0.5"
+    min(x) = x
+    max(x) = x
```
Comment (tkelman) [inline src/intervals/arithmetic.jl]: since you've imported these, this is type piracy and could change the behavior of other unrelated code

File: src/intervals/rounding.jl [hunk 11]
```diff
@@ -0,0 +1,75 @@
+# Define rounded versions of elementary functions
```
Comment (tkelman) [inline src/intervals/rounding.jl]: didn't you have PR's to add these to base?

Comment (dpsanders) [inline src/intervals/rounding.jl]: No, just for `BigFloat` constructors.

File: src/intervals/arithmetic.jl (hunk 10 above)
Comment (dpsanders) [inline src/intervals/arithmetic.jl]: These are just not defined on 0.4. Do you mean that I should change this to `VERSION <= v"0.4"`?

File: src/intervals/arithmetic.jl (hunk 9 above)
Comment (dpsanders) [inline src/intervals/arithmetic.jl]: You mean `isdefined(:Iterators)`? I don't know how to find the exact cutoff where this change was made. In any case, if it's a dev version isn't the expectation that people will keep updating it to the latest version?

File: src/intervals/arithmetic.jl (hunk 10 above)
Comment (tkelman) [inline src/intervals/arithmetic.jl]: No, I mean do you need to be extending a base method on base types? If these are backporting something that worked in later versions of Julia, is it available in Compat?

File: src/intervals/arithmetic.jl (hunk 9 above)
Comment (tkelman) [inline src/intervals/arithmetic.jl]: There should be a PR linked to in NEWS or otherwise not too hard to find, and there's the `contrib/commit-name.sh` script you can use

File: src/intervals/arithmetic.jl (hunk 10 above)
Comment (dpsanders) [inline src/intervals/arithmetic.jl]: These are indeed backporting something that is in 0.5 onwards. It is not available in Compat, but probably should be. https://github.com/JuliaLang/Compat.jl/issues/313

Comment (dpsanders) [inline src/intervals/arithmetic.jl]: Updated in master to use the correct precise versions, thanks.
